### PR TITLE
Add query clause for inner joins

### PR DESCRIPTION
**Pull request by lasinicl**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/22382

```
DeptPerson[] deptPersonList =
       from var person in personList
       join Department dept in deptList
       on person.id equals dept.id
       select {
           fname : person.fname,
           lname : person.lname,
           dept : "Eng"
       };
```

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/JoinClauseNode.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    boolean setDeclaredWithVar();
+
+    boolean isDeclaredWithVar();
+
+    VariableDefinitionNode getVariableDefinitionNode();
+
+    void setVariableDefinitionNode(VariableDefinitionNode variableDefinitionNode);
+
+    OnClauseNode getOnClauseNode();
```
Comment (grainier) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/clauses/JoinClauseNode.java]: Why we keep OnClauseNode here? It's also a top-level node, right? Similar to where clause.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 2]
```diff
@@ -282,6 +283,8 @@
 
     private Stack<BLangNode> queryClauseStack = new Stack<>();
 
+    private Stack<BLangOnClause> onClauseNodeStack = new Stack<>();
```
Comment (grainier) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Don't use this. Use existing queryClauseStack.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 3]
```diff
@@ -2017,12 +2021,52 @@ private void addFromClause(DiagnosticPos pos, Set<Whitespace> ws,
...
+        queryClauseStack.push(joinClause);
+    }
+
     void createOnClause(DiagnosticPos pos, Set<Whitespace> ws) {
         BLangOnClause onClause = (BLangOnClause) TreeBuilder.createOnClauseNode();
         onClause.addWS(ws);
         onClause.pos = pos;
         onClause.expression = (BLangExpression) this.exprNodeStack.pop();
-        queryClauseStack.push(onClause);
+        onClauseNodeStack.push(onClause);
```
Comment (grainier) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Why it's put to onClauseNodeStack?

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 4]
```diff
@@ -918,15 +923,15 @@ doClause
     ;
 
 queryPipeline
-    :   fromClause (fromClause | letClause | whereClause)*
+    :   fromClause ((fromClause | letClause | whereClause )* | (joinClause onClause)?)
```
Comment (grainier) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]: Is onClause mandatory for a join?

File: tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+   int id;
+   string fname;
+   string lname;
+|};
+
+type DeptPersonValue record {|
+    DeptPerson value;
+|};
+
+function getDeptPersonValue((record {| DeptPerson value; |}|error?)|(record {| DeptPerson value; |}?) returnedVal) returns DeptPersonValue? {
```
Comment (grainier) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal]: line exceeds 120 char limit

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 (hunk 4 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]: isn't it?
https://github.com/ballerina-platform/ballerina-lang/issues/22382
it contains the join condition

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/clauses/BLangJoinClause.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package org.wso2.ballerinalang.compiler.tree.clauses;
```
Comment (mohanvive) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/clauses/BLangJoinClause.java:1]: Whether license header in a later PR?

File: tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal [hunk 7]
```diff
@@ -0,0 +1,92 @@
+type DeptPerson record {|
```
Comment (mohanvive) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal:1]: Let's add license header for test files as well. There was a discussion and decided to add license header for test files.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/clauses/BLangJoinClause.java (hunk 6 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/clauses/BLangJoinClause.java:1]: Yes. The license header has been added.

File: tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal (hunk 7 above)
Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal:1]: Added in https://github.com/ballerina-platform/ballerina-lang/pull/27492/commits/03f53110add3cd9ced674ea1fb58213d849bb898

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 8]
```diff
@@ -1941,6 +1943,7 @@ void createQueryExpr(DiagnosticPos pos, Set<Whitespace> ws, boolean isTable, boo
         Collections.reverse(queryClauseStack);
         while (queryClauseStack.size() > 0) {
             queryExpr.addQueryClause(queryClauseStack.pop());
+
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:1946]: unnecessary new line

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 9]
```diff
@@ -1949,72 +1952,99 @@ void startFromClause() {
...
-                               VariableDefinitionNode variableDefinitionNode, boolean isDeclaredWithVar) {
-        BLangFromClause fromClause = (BLangFromClause) TreeBuilder.createFromClauseNode();
-        fromClause.addWS(ws);
-        fromClause.pos = pos;
-        markVariableAsFinal((BLangVariable) variableDefinitionNode.getVariable());
-        fromClause.setVariableDefinitionNode(variableDefinitionNode);
-        fromClause.setCollection(this.exprNodeStack.pop());
-        fromClause.isDeclaredWithVar = isDeclaredWithVar;
-        queryClauseStack.push(fromClause);
+    private void addClause(DiagnosticPos pos, Set<Whitespace> ws,
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:2021]: signature not formatted correctly

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 10]
```diff
@@ -1949,72 +1952,99 @@ void startFromClause() {
...
+                               VariableDefinitionNode variableDefinitionNode, boolean isDeclaredWithVar,
+                           boolean isFromClause) {
+        if (isFromClause) {
+            BLangFromClause fromClause = (BLangFromClause) TreeBuilder.createFromClauseNode();
+            fromClause.addWS(ws);
+            fromClause.pos = pos;
+            markVariableAsFinal((BLangVariable) variableDefinitionNode.getVariable());
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:2028]: this is common for if/else

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 11]
```diff
@@ -1949,72 +1952,99 @@ void startFromClause() {
...
+            BLangJoinClause joinClause = (BLangJoinClause) TreeBuilder.createJoinClauseNode();
+            joinClause.addWS(ws);
+            joinClause.pos = pos;
+            markVariableAsFinal((BLangVariable) variableDefinitionNode.getVariable());
+            joinClause.setVariableDefinitionNode(variableDefinitionNode);
+            joinClause.isDeclaredWithVar = isDeclaredWithVar;
+            joinClause.setCollection(this.exprNodeStack.pop());
+            queryClauseStack.push(joinClause);
+        }
+
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:2043]: unnecessary new line

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 12]
```diff
@@ -1311,6 +1312,20 @@ public void visit(BLangFromClause source) {
         clone.nillableResultType = source.nillableResultType;
     }
 
+    @Override
+    public void visit(BLangJoinClause source) {
+
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:1317]: unnecessary new line at beginning and end. check other places as well

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+        BMap<String, BValue> deptPerson1 = (BMap<String, BValue>) values[0];
+        BMap<String, BValue> deptPerson2 = (BMap<String, BValue>) values[1];
+
+        Assert.assertEquals(deptPerson1.get("fname").stringValue(), "Alex");
+        Assert.assertEquals(deptPerson1.get("lname").stringValue(), "George");
+        Assert.assertEquals(deptPerson2.get("dept").stringValue(), "Eng");
+    }
+
+
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java:57]: unnecessary new line

File: tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+function getDeptPersonValue((record {| DeptPerson value; |}|error?)|(record {| DeptPerson value; |}?) returnedVal)
+returns DeptPersonValue? {
+    var result = returnedVal;
+    if (result is DeptPersonValue) {
+        return result;
+    } else {
+        return ();
+    }
+}
+
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal:31]: unnecessary new line

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 9 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:2021]: `BLangPackageBuilder.java` is no longer used

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 10 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:2028]: `BLangPackageBuilder.java` is no longer used

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 11 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:2043]: `BLangPackageBuilder.java` is no longer used

File: tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal (hunk 14 above)
Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/join-clause.bal:31]: This already fixed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java (hunk 13 above)
Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java:57]: This is already fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java (hunk 12 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:1317]: Fixed in b0e02e2

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 8 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java:1946]: `BLangPackageBuilder` is no longer used

**Pull request merged**